feat: fog of war

amitnos123 opened the pull request:
Most commits are me trying to fix conflicts.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Enhanced fog-of-war visuals with detailed shaders, smooth fade effects, and dynamic environmental elements.
  - Introduced dynamic 2D vision circles to clearly display unit sight ranges.
  - Added a new `FogOfWarManager` to manage fog mechanics effectively.
  - New `BaseComponent` and `BaseManager` classes introduced for improved modularity.
  - Updated game scenes with repositioned entities and new fog management features.

- **Improvements**
  - Refined command processing and team validation for a more stable and responsive gameplay experience.
  - Introduced new visibility and selection components for improved unit interaction.
  - Added new variables and methods for better control over fog size, color, and visibility settings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rluders reviewed (changes requested):
Thanks for your PR! I noticed that the project structure has been reverted to a state prior to the implementation of the feature-based structure. Along with that, some old code has also been reintroduced. It might be worth reviewing the changes again and comparing them with the current structure in the main branch to ensure everything aligns as intended.

Additionally, I would recommend referring to issue #25, which outlines the requirements and suggested organization for the "FoW feature." This could help ensure consistency with the intended design.

Let me know if you need any help!

rluders commented on addons/rts_framework/shaders/detailed_fog_of_war.gdshader:
I guess that we can remove the commented part?

amitnos123 replied:
Yes
My bad

rluders commented on addons/rts_framework/scripts/managers/fog_of_war_manager.gd:
Move to feature based folder as #25

rluders replied:
@amitnos123 this one also needs to be moved inside the `features/vision` folder

rluders commented on addons/rts_framework/scripts/managers/base_manager.gd:
Move to `core/`

rluders commented on addons/rts_framework/scripts/components/queue_component.gd:
You can remove this one, since we already have it inside the `features/construction`

rluders commented on addons/rts_framework/scripts/components/base_component.gd:
Move to `core/`

rluders commented on addons/rts_framework/scenes/fog_of_war_manager.tscn:
Move to the feature folder as #25

rluders commented on addons/rts_framework/scenes/fog_of_war_manager.gdshader:
Move to the feature folder as #25

rluders commented on addons/rts_framework/assets/3D/visibility_shape_3d.tscn:
Move to the feature folder as #25

rluders commented on addons/rts_framework/assets/2D/DynamicCircle2D/DynamicCircle2D.tscn:
Move to the feature folder as #25

rluders commented on addons/rts_framework/assets/2D/DynamicCircle2D/DynamicCircle2D.gd:
Move to the feature folder as #25

rluders commented on addons/rts_framework/scripts/entities/unit_entity.gd:
I think that this has some old code that isn't in use anymore. Probably when you rebase it you brought it back. And this file, was also moved to `addons/rts_framework/entities`

rluders commented on demo/scenes/entities/units/soldier/soldier.gd:
```diff
@@ -0,0 +1,9 @@
+extends UnitEntity
+
+var sight_range = 10
```
It is overwriting the exported property. It would be probably nice to document it.

rluders commented on demo/scenes/entities/units/soldier/soldier.gd:
```diff
@@ -0,0 +1,9 @@
+func _ready() -> void:
+	super()
+
+func is_revealing() -> bool:
+	return true
```
I don' t think that you need to redeclare the `is_revealing`  method here, since it is doing the same from the base class.

rluders commented on demo/scenes/entities/units/soldier/states/attack.gd:
```diff
@@ -0,0 +1,10 @@
+extends State
+
+func enter() -> void:
```
I added some `params` the enter method signature at the state, so you need to update it here:
 
`func enter(params: Dictionaty) -> void:`

rluders commented on demo/scenes/entities/units/soldier/states/build.gd:
I think that a solder cannot perform `build` actions, so it may not need this state.

rluders commented on demo/scenes/entities/units/soldier/states/collect.gd:
It is the same thing here. Can it collect things? I would leave it to some collector unit, mining unit, worker unit...

rluders commented on demo/scenes/entities/units/soldier/states/collect.gd:
```diff
@@ -0,0 +1,10 @@
+extends State
+
+func enter() -> void:
```
I added some `params`  the enter method signature at the state, so you need to update it here:

`func enter(params: Dictionaty) -> void:`

rluders commented on demo/scenes/entities/units/soldier/states/idle.gd:
```diff
@@ -0,0 +1,7 @@
+extends State
+
+func enter() -> void:
```
I added some `params`  the enter method signature at the state, so you need to update it here:

`func enter(params: Dictionaty) -> void:`

rluders commented on demo/scenes/entities/units/soldier/states/move.gd:
This is waaaay to different from the one that we have at the `main`. I think this is probably not correct.

amitnos123 replied:
Most likely yes
Thank you for the PR.

amitnos123 commented:
I tried to fix it according to the PR.
Thank you for the clear PR!

rluders commented on addons/rts_framework/features/vision/DynamicCircle2D.gd:
Maybe we could rename it to `dynamic_circle_2d.gd`

amitnos123 replied:
Yes
The name from [Open RTS by Lampe Games](https://lampe-games.itch.io/open-rts)
My bad not correcting the names

rluders commented on addons/rts_framework/features/vision/DynamicCircle2D.tscn:
Maybe we could rename it to `dynamic_circle_2d.tscn`

rluders reviewed (changes requested):
@amitnos123 Great. I think that there are still some comments to resolve. Could you please check it?

rluders commented on demo/scenes/entities/units/soldier/soldier.gd:
I noticed that your branch contains soldier.gd and soldier.tscn inside demo/scenes/entities/units/soldier/, while in the main branch, the correct version seems to be the one located directly in demo/scenes/entities/units/. Which one is the intended version? I don’t mind adopting a folder-per-unit structure, but we should ensure consistency.

Similarly, the states/ folder exists both inside soldier/states/ and in units/states/. Could you clarify which structure we should follow?

amitnos123 replied:
I deleted the directory `demo/scenes/entities/units/soldier/` on my branch.
I had 2 `soldier.tscn`  files.
one inside `demo/scenes/entities/units/`
and one inside `demo/scenes/entities/units/soldier/`

I looked into it and found I didn't use `demo/scenes/entities/units/soldier/soldier.tscn`
I found it by using `View Owners...` in the Godot editor.

rluders reviewed (changes requested):
It is getting somewhere. I think that we still have some space to improve.

- After I reveal the terrain I can't see my second unit there;
- It looks like just one unit is revealing the FoV;
- Camera position looks weird - did you moved it?
- `VisibleField` has an warning - missing `CollisionShape`?
- The structure is way too complicated - maybe we need to make a scene for it?
![Screenshot From 2025-02-17 19-11-58](https://github.com/user-attachments/assets/6438bd36-9c61-48d9-ae2e-3730658f0a12)
- Also the `CombinedViewport` looks weird.
![Screenshot From 2025-02-17 19-21-27](https://github.com/user-attachments/assets/a75921d3-0a48-4fe2-bc36-3b2634ad4726)

amitnos123 commented:
- `After I reveal the terrain I can't see my second unit there;` - When unit exits `VisibilityField` (Area3D), trigger the function `_on_visibility_field_body_exited` at `fog_of_war_manager.gd` line 102. It turn the visibility off on the body that exit the area. The idea is trigger when a body leaves the vision area. 
- `It looks like just one unit is revealing the FoV;` - Every unit that is part of group `revealed_units`  and function `is_revealing()` return true, it reveal FoW
-  > Camera position looks weird - did you moved it?
 I moved the units and the map. I didn't find a way to do it with the subviewports.
- `VisibleField` warning is from missing a shape. The shapes are added when it sync the units, same when it create the white circule 
-   > The structure is way too complicated - maybe we need to make a scene for it?

  yes

- `CombinedViewport` is to create a  shader for the FoW

amitnos123 commented:
I moved `addons/rts_framework/scripts/managers/fog_of_war_manager.gd`  to `features/vision` directory

rluders commented:
I don't know... it still looks a lot of weird for me, I mean, the code and the structure seems good, but visually at the scene it seems to not look great, and it is also hard to enable/disable it on the editor side.

It also has some weird border effect at the map limit, and when I try to add more units to the map, at its corners, it was not working as expected anymore.

amitnos123 commented:
> I don't know... it still looks a lot of weird for me, I mean, the code and the structure seems good, but visually at the scene it seems to not look great, and it is also hard to enable/disable it on the editor side.

Should I add a debug flag to turn FoW (fog of war) on\off?
I thought of doing a bool export which does it.
In general to hide the FoW, you hide the node `ScreenOverlay` (MeshInstance3D).
Then I can do via editor code which only run in the editor to hide the Fow.

> It also has some weird border effect at the map limit, and when I try to add more units to the map, at its corners, it was not working as expected anymore.

May you upload a picture to explain better and for reference.

amitnos123 commented:
I added to `FogOfWarManager` debug values, which run in the editor.
![2025-02-18_19-07-14](https://github.com/user-attachments/assets/09fb7fec-9516-48e0-8dae-cfd57aa009ce)
![2025-02-18_19-07-05](https://github.com/user-attachments/assets/ef920718-59b0-4c3a-b0f4-caa33aa57f96)
![2025-02-18_19-06-54](https://github.com/user-attachments/assets/8ae9256b-873e-4f79-b7a0-fd44cd28cb4b)
![2025-02-18_19-06-41](https://github.com/user-attachments/assets/32563e3e-07fb-40a9-b5a8-56789053b47a)

amitnos123 commented:
> It also has some weird border effect at the map limit, and when I try to add more units to the map, at its corners, it was not working as expected anymore.

Is this what you mean?

![image](https://github.com/user-attachments/assets/a959f7cd-15dc-4a61-a89a-0e41e3d570b4)

amitnos123 commented:
I added all the values I thought were relevant to `fog_of_war_manager` node itself.
I also added placeholders with TODO for description of all the values
![image](https://github.com/user-attachments/assets/892abb0e-9f06-4c72-a18a-42227964d795)

amitnos123 commented:
I forgot to explain why I chose Area3D to keep track of what the player sees and what he doesn't.
According to [this Reddit post](https://www.reddit.com/r/godot/comments/190j21z/distance_vs_aabb_vs_area3d_for_hundreds_of_checks/), which they compared distance, AABB & Area3D, to check for insight. The most effective why, out of the 3, is Area3D.

amitnos123 commented:
@rluders Is the required change something I missed?

rluders commented:
@amitnos123 I'm a little bit busy this week. I'll check it as ASAP.

amitnos123 commented:
> @amitnos123 I'm a little bit busy this week. I'll check it as ASAP.

It's fine. You don't need to rush.
RL takes priority.
It was strange for me, not to hear from you here.
Hence the message.

Hope your week will be easier!

rluders commented:
@amitnos123 I think that I will have some time to review it today or later Tuesday this week. Don't give up on me. :smile:

amitnos123 commented:
> @amitnos123 I think that I will have some time to review it today or later Tuesday this week. Don't give up on me. 😄

Don't worry, be happy.
I have a lot of free time, between jobs then you know.

I'm looking at other issues and thinking how to implement them.

amitnos123 commented:
I will update to Godot 4.4 after merging this

rluders commented:
Here you can see some of the issues:

https://github.com/user-attachments/assets/ee89a0e3-f01a-4d31-a22b-07128291ae55

amitnos123 commented:
About how it looks, it shaders parameters
The `Texture Units per World Unit` and `Outer Margin for Fade Out` change
About your unit disappearing, it's because you don't see them.
They are outside of your vision. If the unit also disperse fog, you will see them. 

![image](https://github.com/user-attachments/assets/0a3ce278-49d9-45f7-b2dc-1df377f6df09)

rluders commented:
Yes. This is fine. The main issue is that the fog isn't getting the size of the terrain. It should probably get it automatically. Also, the minimap seems to be rectangular, I'm pretty sure that it should be square.

Aside from that, it looks good.

amitnos123 commented:
The `Minimap` isn't a minimap, but a debug tool.
I would try to change it to a square.

About automation of size, I am thinking of doing an export value of the terrain node, which will change its size.
Is it a good idea?

amitnos123 commented:
The debug texture view (minimap-like thingy) is a square.
I added an ability to change it's size
It looks rectangular because of the window size itself.

About size, I did that it update in the editor the fog size to the mesh node size that is given
![image](https://github.com/user-attachments/assets/7ab62203-2cd1-4461-92ed-973377c96f24)

amitnos123 commented on addons/rts_framework/features/vision/dynamic_circle_2d.gd:
```diff
@@ -0,0 +1,9 @@
+@tool
+extends Node2D
+
+@export var color : Color = Color.WHITE
+@export var radius : int = 15
+
+
+func _draw():
+	draw_circle(Vector2(0, 0), radius, color)
```
Added

amitnos123 commented on addons/rts_framework/features/vision/detailed_fog_of_war.gdshader:
```diff
@@ -0,0 +1,116 @@
+      max_distance_from_boundary = max(
+          max_distance_from_boundary, world_position.z - world_visibility_texture_size_scaled.y);
+      float alpha_from_margin =
+          smoothstep(0.0, outer_margin_for_fade_out, max_distance_from_boundary) * 0.2 + 0.8;
```
Added

rluders commented:
This is already getting way to big. If we found any issue we still can solve/improve it later. This was a really good job and a must have feature. Thanks, @amitnos123.

amitnos123 commented:
@rluders Thank you for starting the framework
I miss C&C RTS games.
